### PR TITLE
FM2-132 : Implement update functionality to allergy resource

### PR DESCRIPTION
**Pull request by jecihjoy**
## Issue I worked on

see https://issues.openmrs.org/browse/FM2-132

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**.

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirAllergyIntoleranceServiceImplTest.java [hunk 1]
```diff
@@ -276,4 +279,47 @@ public void saveAllergy_shouldSaveNewAllergy() {
...
+	@Test(expected = InvalidRequestException.class)
+	public void updateAllergy_shouldThrowInvalidRequestExceptionIfIdIsNull() {
+		AllergyIntolerance allergyIntolerance = new AllergyIntolerance();
+		allergyIntolerance.setId(ALLERGY_UUID);
+		
+		AllergyIntolerance result = service.update(null, allergyIntolerance);
+	}
+	
+	@Test(expected = InvalidRequestException.class)
+	public void updateAllergy_shouldThrowInvalidRequestException() {
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirAllergyIntoleranceServiceImplTest.java:291]: I'd prefer `updateAllergy_shouldThrowInvalidRequestExceptionWhenResourceIdDoesntMatchUrl` or something so it's clear what's being tested (yes its verbose)

File: api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirAllergyIntoleranceServiceImplTest.java [hunk 2]
```diff
@@ -276,4 +279,47 @@ public void saveAllergy_shouldSaveNewAllergy() {
...
+		AllergyIntolerance allergyIntolerance = new AllergyIntolerance();
+		allergyIntolerance.setId(ALLERGY_UUID);
+		
+		AllergyIntolerance result = service.update(WRONG_ALLERGY_UUID, allergyIntolerance);
+	}
+	
+	@Test(expected = MethodNotAllowedException.class)
+	public void updateAllergy_shouldThrowMethodNotAllowedException() {
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirAllergyIntoleranceServiceImplTest.java:299]: Why should it throw a method not allowed exception?

File: omod/src/main/java/org/openmrs/module/fhir2/providers/AllergyIntoleranceFhirResourceProvider.java [hunk 3]
```diff
@@ -93,4 +94,13 @@ public MethodOutcome createAllergy(@ResourceParam AllergyIntolerance allergy) {
 		return FhirServerUtils.buildCreate(fhirAllergyIntoleranceService.create(allergy));
 	}
 	
+	@Update
+	@SuppressWarnings("unused")
+	public MethodOutcome updateAllergy(@ResourceParam AllergyIntolerance allergy, @IdParam IdType id) {
+		if (id != null) {
+			allergy.setId(id.getIdPart());
+		}
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/fhir2/providers/AllergyIntoleranceFhirResourceProvider.java:102]: I don't love changing the resource id automatically. This allows easy errors (posting an updated resource to the wrong url) to potentially overwrite important information. I would just as soon not have this check here.

**Comment by varung-31** [comment]
@ibacher A similar task has been completed in #212 
 I think we can close this pull request.

**Comment by ibacher** [comment]
As @varung-31, we've implemented this elsewhere.

**Pull request closed without merging**